File: 309. Best Time to Buy and Sell Stock with Cooldown/309. Best Time to Buy and Sell Stock with Cooldown bottom-up.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int[][] dp = new int[n + 1][2];

        for (int i = n - 1; i >= 0; i--) {
            for (int holding = 0; holding <= 1; holding++) {
                int do_nothing = dp[i + 1][holding];
                int do_something = 0;
                if (holding == 1) {
                    // sell stock
                    do_something = i < n - 1 ? prices[i] + dp[i + 2][0] : prices[i];
                } else {
                    // buy stock
                    do_something = -prices[i] + dp[i + 1][1];
                }
                dp[i][holding] = Math.max(do_nothing, do_something);
            }
        }
        return dp[0][0];
    }
}

File: 309. Best Time to Buy and Sell Stock with Cooldown/309. Best Time to Buy and Sell Stock with Cooldown bottom-up.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        dp = [[0]*2 for _ in range(n+1)]
        for i in range(n-1, -1, -1):
            for holding in range(2):
                do_nothing = dp[i+1][holding]
                if holding:
                    # sell stock
                    do_something = prices[i] + \
                        dp[i+2][0] if i < n-1 else prices[i]
                else:
                    # buy stock
                    do_something = -prices[i] + dp[i+1][1]
                dp[i][holding] = max(do_nothing, do_something)
        return dp[0][0]
